website/blog/2023-04-04-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+*Figure 6: HoodieTableInputFormat implementations.*
+
+For `_rt` tables, input format is registered as `HoodieParquetRealtimeInputFormat`. Hence when `inputFormat.getSplits()` is called, it calls `HoodieParquetInputFormatBase.getSplits()` which delegates the call to `HoodieTableInputFormat` which is implemented as `HoodieMergeOnReadTableInputFormat.getSplits()`. This makes a call to `FileInputFormat.getSplits()`.
+
+Now refer to Figure 1, `listStatus()` is called which is overridden in `HoodieCopyOnWriteTableInputFormat.listStatus()`. This method in turn calls `listStatusForIncrementalMode()` and `listStatusForSnapshotMode()`. This will return `RealtimeFileStatus`. `RealtimeFileStatus` overrides `getPath()` method which returns `HoodieRealtimePath`. Please refer to the below figures -  
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/list-status-mor.png)
+*Figure 7: listStatus() call for MoR table type returns RealtimeFileStatus.*
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-mor.png)
+
+*Figure 8: makeSplit() internally called from getSplits() for MoR table type returns HoodieRealtimeFileSplit.*
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-path-mor.png)
+
+*Figure 9: getPath() is overridden in RealtimeFileStatus and it returns HoodieRealtimePath.*
+
+Now that real time splits got generated, we need a way of reading the records from these special splits by merging the base and delta files. `GenericHiveRecordCursorProvider` calls createRecordReader which ultimately calls `inputFormat.getRecordReader()`. This returns `RealtimeCompactedRecordReader` by default. Please refer to the below figure - 
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/real-time-record-reader.png)
+*Figure 10: Real time record reader.*
+
+
+#### Read Optimized Queries
+
+These queries will occur from `<table_name>_ro` table registered in HMS and it has HoodieParquetInputFormat as the input format. These queries will only read the base parquet files, hence the name read optimized. The execution of these queries is exactly similar to that for CoW table type. Please refer to the next section for more details.
+
+### Copy on Write table type
+
+The layout of this table type only consists of parquet files. This makes use of the HudiDirectoryLister class for listing relevant data files. HudiDirectoryLister is a singleton object and gets created once every query involving hudi CoW table. This class creates the hudi timeline by scanning the .hoodie folder, then creates a file system view on top of this timeline. FileSystemView is a logical representation of the hudi storage layout which binds the timeline and data files together.
+
+Unlike HadoopDirectoryLister, this lister does not make use of the file listing cache since the hudi table is assumed to be ever evolving and hence the partitions are not sealed. Once the splits are created, native ParquetPageSource is used for rendering the results.
```

Comment:
```suggestion
Unlike `HadoopDirectoryLister`, this lister does not make use of the file listing cache since the hudi table is assumed to be ever-evolving and hence the partitions are not sealed. Once the splits are created, native `ParquetPageSource` is used for rendering the results.
```